starlette/testclient.py

```diff
@@ -152,11 +177,40 @@ async def _asgi_receive(self) -> Message:
     async def _asgi_send(self, message: Message) -> None:
         self._send_queue.put(message)
 
-    def _raise_on_close(self, message: Message) -> None:
+    def _handle_response(self, message: Message) -> None:
+        status_code: int = message["status"]
+        headers: typing.List[typing.Tuple[bytes, bytes]] = message["headers"]
+        body = []
+        while True:
+            message = self.receive()
+            assert message["type"] == "websocket.http.response.body"
+            body.append(message["body"])
+            if not message.get("more_body", False):
+                break
+        raise WebSocketReject(
+            response_status=status_code,
+            response_headers=headers,
+            response_body=b"".join(body),
+        )
+
+    def _raise_on_close(
+        self, message: Message, reject: typing.Optional[bool] = None
+    ) -> None:
         if message["type"] == "websocket.close":
-            raise WebSocketDisconnect(
-                message.get("code", 1000), message.get("reason", "")
-            )
+            if not reject:
+                raise WebSocketDisconnect(
+                    message.get("code", 1000), message.get("reason", "")
+                )
+            else:
+                # A webserver which gets a "close" before "accept"
+                # will return a 403 response.  It may or may not do anything
+                # with the close code and reason.  We store it in this exception
+                # for test introspection.
+                raise WebSocketReject(
+                    response_status=403,
+                    close_code=message.get("code", 1000),
+                    close_reason=message.get("reason"),
+                )
```

Comment:
This is actually a change that doesn't need to be in this PR. This is actually "fixing" the current state, since the client would actually receive a 403, instead of the code, and reason on the WebSocket.

Please remove this from the PR. Let's discuss this after.

Reply:
Okay